src/source.cpp

```diff
@@ -0,0 +1,379 @@
+      fatal_error("Source energy below range of energies of at least "
+                  "one cross section table");
+    }
+  }
+
+  while (true) {
+    // Sample energy spectrum
+    site.E = energy_->sample();
+
+    // Resample if energy falls outside minimum or maximum particle energy
+    if (site.E < energy_max[p-1] && site.E > energy_min[p-1]) break;
+  }
+
+  // Set delayed group
+  site.delayed_group = 0;
+
+  return site;
+}
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+void initialize_source()
+{
+  write_message("Initializing source particles...", 5);
+
+  // Get pointer to source bank
+  Bank* source_bank;
+  int64_t n;
+  openmc_source_bank(&source_bank, &n);
+
+  if (path_source != "") {
+    // Read the source from a binary file instead of sampling from some
+    // assumed source distribution
+
+    std::stringstream msg;
+    msg << "Reading source file from " << path_source << "...";
+    write_message(msg, 6);
+
+    // Open the binary file
+    hid_t file_id = file_open(path_source, 'r', true);
+
+    // Read the file type
+    std::string filetype;
+    read_attribute(file_id, "filetype", filetype);
+
+    // Check to make sure this is a source file
+    if (filetype != "source" && filetype != "statepoint") {
+      fatal_error("Specified starting source file not a source file type.");
+    }
+
+    // Read in the source bank
+    read_source_bank(file_id, work_index.data(), source_bank);
+
+    // Close file
+    file_close(file_id);
+
+  } else {
+    // Generation source sites from specified distribution in user input
+    for (int64_t i = 0; i < openmc_work; ++i) {
+      // initialize random number seed
+      int64_t id = openmc_total_gen*n_particles + work_index[openmc::mpi::rank] + i + 1;
+      set_particle_seed(id);
+
+      // sample external source distribution
+      source_bank[i] = sample_external_source();
+    }
+  }
+
+  // Write out initial source
+  if (openmc_write_initial_source) {
+    write_message("Writing out initial source...", 5);
+    std::string filename = path_output + "initial_source.h5";
+    hid_t file_id = file_open(filename, 'w', true);
+    write_source_bank(file_id, work_index.data(), source_bank);
+    file_close(file_id);
+  }
+}
+
+extern "C" double* rev_energy_bins_ptr();
+
+Bank sample_external_source()
+{
+  // Set the random number generator to the source stream.
+  prn_set_stream(STREAM_SOURCE);
+
+  // Determine total source strength
+  double total_strength = 0.0;
+  for (auto& s : external_sources)
+    total_strength += s.strength();
+
+  // Sample from among multiple source distributions
+  auto n_source = external_sources.size();
+  int i = 0;
+  if (n_source > 1) {
+    double xi = prn()*total_strength;
+    double c = 0.0;
+    for (i = 0; i < external_sources.size(); ++i) {
```

Comment:
I found it slightly confusing that `i` was initialized both here and above.  Maybe just
```C++
for (; i < external_sources.size(); ++i) {
```
for clarity.  (Or just declare `int i;` above)